dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -236,16 +236,64 @@ impl<'a, 'tcx: 'a> CollectFunctionInstrumentationPoints<'a, 'tcx> {
         find_instrumentation_def(self.tcx, self.runtime_crate_did, name)
     }
 
-    fn should_instrument(&self, local: Local, context: PlaceContext) -> bool {
-        let is_unsafe_ptr = self.body.local_decls[local].ty.is_unsafe_ptr();
-        is_unsafe_ptr && context.is_use()
-    }
-
     fn func_hash(&self) -> mir_loc::DefPathHash {
         self.tcx.def_path_hash(self.body.source.def_id()).convert()
     }
 }
 
+fn is_shared_or_unsafe_ptr(ty: &TyS) -> bool {
+    ty.is_unsafe_ptr() || (ty.is_region_ptr() && !ty.is_mutable_ptr())
+}
+
+fn is_region_or_unsafe_ptr(ty: &TyS) -> bool {
+    ty.is_unsafe_ptr() || ty.is_region_ptr()
+}
+
+fn has_outer_deref(p: &Place) -> bool {
+    matches!(
+        p.iter_projections().last(),
+        Some((_, ProjectionElem::Deref))
+    )
+}
+
+fn sans_last_projection<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
+    let Place { local, projection } = *p;
+    projection.split_last().map(|(_last, rest)| Place {
+        local,
+        projection: tcx.intern_place_elems(rest),
+    })
+}
+
+/// Get the inner-most dereferenced [`Place`].
+fn strip_all_deref<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Place<'tcx> {
+    let mut base_dest = *p;
+    let mut place_ref = base_dest.as_ref();
+    while let Some((cur_ref, proj)) = place_ref.last_projection() {
+        if let ProjectionElem::Deref = proj {
+            base_dest = Place {
+                local: cur_ref.local,
+                projection: tcx.intern_place_elems(cur_ref.projection),
```

Comment:
Nit: This will intern multiple times if there are multiple derefs, right? Might be better to keep a ref to the inner-most deref and then intern and return that at the end?